In [52]:
import graphlab

In [53]:
data1 = graphlab.SFrame('breast-cancer-wisconsin.csv')
data3 = graphlab.SFrame('ovarian.csv')

Finished parsing file /Users/ayush/Desktop/Machine Learning/Cancer Model Generalized/breast-cancer-wisconsin.csv

Parsing completed. Parsed 100 lines in 0.03523 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[int,int,int,int,int,int,str,int,int,int,int]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Finished parsing file /Users/ayush/Desktop/Machine Learning/Cancer Model Generalized/breast-cancer-wisconsin.csv

Parsing completed. Parsed 699 lines in 0.020988 secs.

Finished parsing file /Users/ayush/Desktop/Machine Learning/Cancer Model Generalized/ovarian.csv

Parsing completed. Parsed 100 lines in 0.040246 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[float,float]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 15154 lines. Lines per second: 330924

Finished parsing file /Users/ayush/Desktop/Machine Learning/Cancer Model Generalized/ovarian.csv

Parsing completed. Parsed 15154 lines in 0.046272 secs.

In [54]:
data1


Columns:
	id	int
	clump_thickness	int
	size_uniformity	int
	shape_uniformity	int
	marginal_adhesion	int
	epithelial_size	int
	bare_nuclei	str
	bland_chromatin	int
	normal_nucleoli	int
	mitoses	int
	class	int

Rows: 699

Data:
+---------+-----------------+-----------------+------------------+-------------------+
|    id   | clump_thickness | size_uniformity | shape_uniformity | marginal_adhesion |
+---------+-----------------+-----------------+------------------+-------------------+
| 1000025 |        5        |        1        |        1         |         1         |
| 1002945 |        5        |        4        |        4         |         5         |
| 1015425 |        3        |        1        |        1         |         1         |
| 1016277 |        6        |        8        |        8         |         1         |
| 1017023 |        4        |        1        |        1         |         3         |
| 1017122 |        8        |        10       |        10        |         8         |
| 1018099 |        1        |        1        |        1         |         1         |
| 1018561 |        2        |        1        |        2         |         1         |
| 1033078 |        2        |        1        |        1         |         1         |
| 1033078 |        4        |        2        |        1         |         1         |
+---------+-----------------+-----------------+------------------+-------------------+
+-----------------+-------------+-----------------+-----------------+---------+
| epithelial_size | bare_nuclei | bland_chromatin | normal_nucleoli | mitoses |
+-----------------+-------------+-----------------+-----------------+---------+
|        2        |      1      |        3        |        1        |    1    |
|        7        |      10     |        3        |        2        |    1    |
|        2        |      2      |        3        |        1        |    1    |
|        3        |      4      |        3        |        7        |    1    |
|        2        |      1      |        3        |        1        |    1    |
|        7        |      10     |        9        |        7        |    1    |
|        2        |      10     |        3        |        1        |    1    |
|        2        |      1      |        3        |        1        |    1    |
|        2        |      1      |        1        |        1        |    5    |
|        2        |      1      |        2        |        1        |    1    |
+-----------------+-------------+-----------------+-----------------+---------+
+-------+
| class |
+-------+
|   2   |
|   2   |
|   2   |
|   2   |
|   2   |
|   4   |
|   2   |
|   2   |
|   2   |
|   2   |
+-------+
[699 rows x 11 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [55]:
graphlab.canvas.set_target('ipynb')
data1.show(view="Summary")

### Visualizing Data (Standard)

In [56]:
data1.show(view="Line Chart", x="clump_thickness", y='id')

In [58]:
data1.show(view="Bar Chart", x="bland_chromatin", y='id')

In [60]:
data1.show(view="BoxWhisker Plot", x="marginal_adhesion", y='id')

In case of standard data, we observe the following trend: The higher the value of the feature (such as bland chromatin, marginal adhesion), the higher the change for cancer malignancy, and in turn increased probability for mortality.

We model the above data into two classification models: ANN and SVM, to asses how different data sets affect model metrics such as efficiency, specificity etc. 

### Visualizing Data (Gene expression)

In [61]:
data3

Columns:
	M/Z	float
	Intensity	float

Rows: 15154

Data:
+----------------+-----------+
|      M/Z       | Intensity |
+----------------+-----------+
| -7.8602611e-05 | 4.1045752 |
| 2.1773576e-07  | 4.1106083 |
| 9.6021472e-05  | 4.0542986 |
| 0.00036601382  | 4.1206637 |
| 0.00081019477  | 4.0221217 |
|  0.0014285643  | 3.9537456 |
|  0.0022211225  | 3.8833585 |
|  0.0031878693  | 3.9698341 |
|  0.0043288047  | 4.0180995 |
|  0.0056439287  | 4.0100553 |
+----------------+-----------+
[15154 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [62]:
data3.show(view="Summary")

In [63]:
data3.show(view="Line Chart", x="M/Z", y="Intensity")

In [65]:
data3.show(view="BoxWhisker Plot", x="M/Z", y="Intensity")

The cancer data for genetic dataset is highly spread and therefore it is difficult to observe a consistent trend among the data entries. However, for binary classification purposes it has been medically proven that an intensity of more than 50 is more likely to engender ovarian/prostate cancer. 

Just like breast cancer, where physical examination reveals features/attributes of data sets. Here, the examination marke is given by the process of CA125, leading to the following data set.

### Creating the artificial neural network model

In [66]:
train_data, test_data = data1.random_split(0.8)
feature_array = ['clump_thickness', 'size_uniformity', 'shape_uniformity', 'marginal_adhesion', 'bland_chromatin', 'mitoses', 'epithelial_size', 'normal_nucleoli']
model_neural = graphlab.neuralnet_classifier.create(train_data, target='class', features=feature_array)

Using network:

### network layers ###
layer[0]: FullConnectionLayer
  init_sigma = 0.01
  init_random = gaussian
  init_bias = 0
  num_hidden_units = 10
layer[1]: SigmoidLayer
layer[2]: FullConnectionLayer
  init_sigma = 0.01
  init_random = gaussian
  init_bias = 0
  num_hidden_units = 2
layer[3]: SoftmaxLayer
### end network layers ###

### network parameters ###
learning_rate = 0.001
momentum = 0.9
### end network parameters ###

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Creating neuralnet using cpu

Training with batch size = 100

+-----------+----------+--------------+-------------------+---------------------+-----------------+

| Iteration | Examples | Elapsed Time | Training-accuracy | Validation-accuracy | Examples/second |

+-----------+----------+--------------+-------------------+---------------------+-----------------+

| 1         | 600      | 0.015295     | 0.633333          | 0.666667            | 39231.066406    |

| 2         | 600      | 0.026860     | 0.633333          | 0.666667            | 54300.093750    |

| 3         | 600      | 0.037755     | 0.633333          | 0.666667            | 55421.531250    |

| 4         | 600      | 0.050197     | 0.633333          | 0.666667            | 48490.484375    |

| 5         | 600      | 0.062592     | 0.633333          | 0.666667            | 48764.730469    |

| 6         | 600      | 0.073847     | 0.633333          | 0.666667            | 53666.167969    |

| 7         | 600      | 0.089570     | 0.633333          | 0.666667            | 38402.070312    |

| 8         | 600      | 0.103875     | 0.633333          | 0.666667            | 42801.320312    |

| 9         | 600      | 0.120344     | 0.633333          | 0.666667            | 36647.769531    |

| 10        | 600      | 0.133563     | 0.633333          | 0.666667            | 45697.117188    |

+-----------+----------+--------------+-------------------+---------------------+-----------------+

In [67]:
model_neural.show(view="Summary")

In [68]:
pred = model_neural.classify(test_data)

In [69]:
pred

Columns:
	row_id	int
	class	int
	probability	float

Rows: 147

Data:
+--------+-------+----------------+
| row_id | class |  probability   |
+--------+-------+----------------+
|   0    |   2   | 0.587500989437 |
|   1    |   2   | 0.585788607597 |
|   2    |   2   | 0.588369131088 |
|   3    |   2   | 0.588559150696 |
|   4    |   2   | 0.584629893303 |
|   5    |   2   | 0.581713438034 |
|   6    |   2   | 0.588684380054 |
|   7    |   2   | 0.585021317005 |
|   8    |   2   | 0.586781799793 |
|   9    |   2   | 0.586582005024 |
+--------+-------+----------------+
[147 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [70]:
eval = model_neural.evaluate(test_data)

In [71]:
eval

{'accuracy': 0.680272102355957, 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 2
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      2       |        2        |  100  |
 |      4       |        2        |   47  |
 +--------------+-----------------+-------+
 [2 rows x 3 columns]}

In [72]:
pred_topk = model_neural.predict_topk(test_data, k=2)

In [73]:
pred_topk

Columns:
	row_id	int
	class	int
	probability	float

Rows: 294

Data:
+--------+-------+----------------+
| row_id | class |  probability   |
+--------+-------+----------------+
|   0    |   2   | 0.587500989437 |
|   0    |   4   | 0.412498980761 |
|   1    |   2   | 0.585788607597 |
|   1    |   4   | 0.414211392403 |
|   2    |   2   | 0.588369131088 |
|   2    |   4   | 0.411630839109 |
|   3    |   2   | 0.588559150696 |
|   3    |   4   | 0.411440849304 |
|   4    |   2   | 0.584629893303 |
|   4    |   4   | 0.415370106697 |
+--------+-------+----------------+
[294 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [74]:
model_neural.show(view="Evaluation")

### Comparing with the support vector machine model.

In [75]:
model_vector = graphlab.svm_classifier.create(train_data, target='class', features=feature_array, validation_set='auto', verbose=True)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



SVM:

--------------------------------------------------------

Number of examples          : 526

Number of classes           : 2

Number of feature columns   : 8

Number of unpacked features : 8

Number of coefficients    : 9

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 1         | 3        | 0.001901  | 0.007287     | 0.351711          | 0.346154            |

| 2         | 7        | 3.000000  | 0.014095     | 0.882129          | 0.807692            |

| 3         | 8        | 3.000000  | 0.015459     | 0.758555          | 0.730769            |

| 4         | 10       | 1.000000  | 0.019766     | 0.952471          | 0.884615            |

| 5         | 11       | 1.000000  | 0.024934     | 0.952471          | 0.884615            |

| 6         | 12       | 1.000000  | 0.029081     | 0.908745          | 0.961538            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

In [76]:
model_vector.show()

In [77]:
prediction_vector = model_vector.predict(test_data)

In [78]:
prediction_vector.show()

In [79]:
eval_ = model_vector.evaluate(test_data)

In [81]:
eval_

{'accuracy': 0.9863945578231292, 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 3
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      4       |        4        |   47  |
 |      2       |        4        |   2   |
 |      2       |        2        |   98  |
 +--------------+-----------------+-------+
 [3 rows x 3 columns], 'f1_score': 0.9791666666666666, 'precision': 0.9591836734693877, 'recall': 1.0}

We see that for a data set with less entries, support vector machine performs better as it translates into a linear model which evaluates better compared to artificial neural network.

## Using above techniques on BUPA Liver Cancer Data set

In [82]:
data2 = graphlab.SFrame('bupa.csv')

Finished parsing file /Users/ayush/Desktop/Machine Learning/Cancer Model Generalized/bupa.csv

Parsing completed. Parsed 100 lines in 0.051531 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[int,int,int,int,int,float,int]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Finished parsing file /Users/ayush/Desktop/Machine Learning/Cancer Model Generalized/bupa.csv

Parsing completed. Parsed 345 lines in 0.01836 secs.

In [83]:
data2

Columns:
	mcv	int
	alkphos	int
	sgpt	int
	sgot	int
	gammagt	int
	drink	float
	class	int

Rows: 345

Data:
+-----+---------+------+------+---------+-------+-------+
| mcv | alkphos | sgpt | sgot | gammagt | drink | class |
+-----+---------+------+------+---------+-------+-------+
|  85 |    92   |  45  |  27  |    31   |  0.0  |   1   |
|  85 |    64   |  59  |  32  |    23   |  0.0  |   2   |
|  86 |    54   |  33  |  16  |    54   |  0.0  |   2   |
|  91 |    78   |  34  |  24  |    36   |  0.0  |   2   |
|  87 |    70   |  12  |  28  |    10   |  0.0  |   2   |
|  98 |    55   |  13  |  17  |    17   |  0.0  |   2   |
|  88 |    62   |  20  |  17  |    9    |  0.5  |   1   |
|  88 |    67   |  21  |  11  |    11   |  0.5  |   1   |
|  92 |    54   |  22  |  20  |    7    |  0.5  |   1   |
|  90 |    60   |  25  |  19  |    5    |  0.5  |   1   |
+-----+---------+------+------+---------+-------+-------+
[345 rows x 7 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [84]:
graphlab.canvas.set_target('ipynb')
data2.show(view="Summary")

In [85]:
data2.show(view="Line Chart", x="sgpt", y='class')

#### Observation: SGPT for patients are higher for malignant ones compared to those of benign ones. Similar trend is observed for other features as well.

In [86]:
data2.show(view="Bar Chart", x="alkphos", y='class')

In [87]:
data2.show(view="BoxWhisker Plot", x="sgot", y='class')

In [88]:
train_data2, test_data2 = data2.random_split(0.8)
feature_array2 = ['alkphos', 'sgot', 'sgpt', 'mcv', 'gammagt', 'drink']
model_neural2 = graphlab.neuralnet_classifier.create(train_data2, target='class', features=feature_array2)

Using network:

### network layers ###
layer[0]: FullConnectionLayer
  init_sigma = 0.01
  init_random = gaussian
  init_bias = 0
  num_hidden_units = 10
layer[1]: SigmoidLayer
layer[2]: FullConnectionLayer
  init_sigma = 0.01
  init_random = gaussian
  init_bias = 0
  num_hidden_units = 2
layer[3]: SoftmaxLayer
### end network layers ###

### network parameters ###
learning_rate = 0.001
momentum = 0.9
### end network parameters ###

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Creating neuralnet using cpu

Training with batch size = 100

+-----------+----------+--------------+-------------------+---------------------+-----------------+

| Iteration | Examples | Elapsed Time | Training-accuracy | Validation-accuracy | Examples/second |

+-----------+----------+--------------+-------------------+---------------------+-----------------+

| 1         | 300      | 0.018581     | 0.526667          | 0.545455            | 16145.625000    |

| 2         | 300      | 0.034315     | 0.526667          | 0.545455            | 19192.541016    |

| 3         | 300      | 0.058216     | 0.526667          | 0.545455            | 12636.714844    |

| 4         | 300      | 0.073227     | 0.526667          | 0.545455            | 20104.707031    |

| 5         | 300      | 0.088322     | 0.526667          | 0.545455            | 20002.417969    |

| 6         | 300      | 0.102057     | 0.526667          | 0.545455            | 21962.865234    |

| 7         | 300      | 0.117226     | 0.526667          | 0.545455            | 19894.878906    |

| 8         | 300      | 0.138899     | 0.526667          | 0.545455            | 14099.475586    |

| 9         | 300      | 0.154736     | 0.526667          | 0.545455            | 19207.623047    |

| 10        | 300      | 0.172051     | 0.526667          | 0.545455            | 17433.500000    |

+-----------+----------+--------------+-------------------+---------------------+-----------------+

In [89]:
model_neural2.show(view="Summary")

In [90]:
pred2 = model_neural2.classify(test_data2)

In [91]:
pred2

Columns:
	row_id	int
	class	int
	probability	float

Rows: 61

Data:
+--------+-------+----------------+
| row_id | class |  probability   |
+--------+-------+----------------+
|   0    |   2   | 0.514295995235 |
|   1    |   2   | 0.514591932297 |
|   2    |   2   | 0.51342612505  |
|   3    |   2   | 0.513011157513 |
|   4    |   2   | 0.51477253437  |
|   5    |   2   | 0.513086438179 |
|   6    |   2   | 0.513864278793 |
|   7    |   2   | 0.513604283333 |
|   8    |   2   | 0.51284044981  |
|   9    |   2   | 0.517160654068 |
+--------+-------+----------------+
[61 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [92]:
eval2 = model_neural2.evaluate(test_data2)

In [93]:
eval2

{'accuracy': 0.6393442749977112, 'confusion_matrix': Columns:
 	target_label	int
 	predicted_label	int
 	count	int
 
 Rows: 2
 
 Data:
 +--------------+-----------------+-------+
 | target_label | predicted_label | count |
 +--------------+-----------------+-------+
 |      1       |        2        |   22  |
 |      2       |        2        |   39  |
 +--------------+-----------------+-------+
 [2 rows x 3 columns]}

We see that neural network struggles to find a good accuracy and precision due to low number of entries in the data set. (345 to be exact.)

#### Modelling with SVM (Liver Cancer)

In [94]:
model_vector2 = graphlab.svm_classifier.create(train_data2, target='class', features=feature_array2, validation_set='auto', verbose=True)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



SVM:

--------------------------------------------------------

Number of examples          : 271

Number of classes           : 2

Number of feature columns   : 6

Number of unpacked features : 6

Number of coefficients    : 7

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy | Validation-accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 1         | 3        | 0.003690  | 0.000971     | 0.571956          | 0.461538            |

| 2         | 5        | 1.000000  | 0.002608     | 0.571956          | 0.461538            |

| 3         | 6        | 1.000000  | 0.007300     | 0.571956          | 0.461538            |

| 4         | 7        | 1.000000  | 0.008384     | 0.575646          | 0.461538            |

| 5         | 8        | 1.000000  | 0.009530     | 0.571956          | 0.461538            |

| 6         | 9        | 1.000000  | 0.010966     | 0.571956          | 0.461538            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

TERMINATED: Iteration limit reached.

This model may not be optimal. To improve it, consider increasing `max_iterations`.

In [95]:
model_vector2.show(view = "Summary")

In [96]:
prediction_2 = model_vector2.predict(test_data2)

In [97]:
prediction_2.show()


In [98]:
evaluation_2 = model_vector2.evaluate(test_data2)

In [99]:
model_vector2.show(view="Evaluation")